src/Features/Core/Portable/ImplementInterface/AbstractImplementInterfaceService.CodeAction.cs

```diff
@@ -476,13 +476,36 @@ private SyntaxNode CreateThroughExpression(SyntaxGenerator factory)
                     // uncommon case and optimize for the common one - in other words, we only apply the cast
                     // in cases where we can unambiguously figure out which interface we are trying to implement.
                     var interfaceBeingImplemented = State.InterfaceTypes.SingleOrDefault();
-                    if ((interfaceBeingImplemented != null) && (!throughMemberType.Equals(interfaceBeingImplemented)))
+                    if ((interfaceBeingImplemented != null))
                     {
-                        through = factory.CastExpression(interfaceBeingImplemented,
-                            through.WithAdditionalAnnotations(Simplifier.Annotation));
+                        if (!throughMemberType.Equals(interfaceBeingImplemented))
+                        {
+                            through = factory.CastExpression(interfaceBeingImplemented,
+                                through.WithAdditionalAnnotations(Simplifier.Annotation));
+
+                            var facts = this.Document.GetLanguageService<ISyntaxFactsService>();
+                            through = facts.Parenthesize(through);
+                        }
+                        else if (!ThroughMember.IsStatic && ThroughMember is IPropertySymbol throughMemberProperty && throughMemberProperty.ExplicitInterfaceImplementations.Any())
+                        {
+                            // If we are implementing through an explicitly implemented property, we need to cast 'this' to
+                            // the explicitly implemented property type before calling the member, as in:
+                            //       ((IA)this).Prop.Member();
+                            //
+                            var explicitlyImplementedProperty = throughMemberProperty.ExplicitInterfaceImplementations[0];
 
-                        var facts = this.Document.GetLanguageService<ISyntaxFactsService>();
-                        through = facts.Parenthesize(through);
+                            var explicitImplementationCast = factory.CastExpression(
+                                explicitlyImplementedProperty.ContainingType,
+                                factory.ThisExpression());
+
+                            var facts = this.Document.GetLanguageService<ISyntaxFactsService>();
```

Comment:
it really shoudn't be necessary.  factory.MemberAccessExpression should parenthesize the left side of the dot.  see: 
https://github.com/dotnet/roslyn/blob/master/src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs#L3561